src/lib/Types/Core.hs

```diff
@@ -295,6 +299,13 @@ data DictExpr (n::S) =
  | SuperclassProj (Atom n) Int  -- (could instantiate here too, but we don't need it for now)
    -- Special case for `Ix (Fin n)`  (TODO: a more general mechanism for built-in classes and instances)
  | IxFin (Atom n)
+   -- Used for dicts defined by top-level functions, which may take extra data parameters
+   -- TODO: consider bundling `(DictType n, [AtomName n])` as a top-level
+   -- binding for some `Name DictNameC` to make the IR smaller.
+   -- TODO: the function atoms should be names. We could enforce that syntactically but then
+   -- we can't derive `SubstE AtomSubstVal`. Maybe we should have a separate name color for
+   -- top function names.
+ | ExplicitMethods (DictType n) [Atom n] [Atom n] -- dict type, names of parameterized method functions, parameters
```

Comment:
Don't we have another way to create an explicit dict? perhaps we should unify those two?

Reply:
Also, just to clarify: names of parameterized method functions here refer to the fact that those `Atom`s are really `AtomName`s?

Reply:
> Don't we have another way to create an explicit dict? perhaps we should unify those two?

Yes, we previously had dicts representable by a newtype-cast of a product of methods. This change removes that version completely. Keeping them non-inlined like this should make the IR smaller.

> Also, just to clarify: names of parameterized method functions here refer to the fact that those Atoms are really AtomNames?

Correct.

Reply:
Right. I know that you removed the newtype, but we still have `Con (ExplicitDict ...)` which seems equivalent to what you have here.

Reply:
I see. It's definitely not equivalent. `Con (ExplicitDict ...)` takes the method as an atom, and only allows single methods, whereas `ExplicitMethods` gives the data-parameterized methods as names, and supplies parameter. But we only use the `Con (ExplicitDict ...)` for a single purpose, which is to supply the `accumBaseMonoid` in `run_accum`, and maybe we can do the same thing by exposing surface syntax for single-method `ExplicitMethods`. I can try it.

Reply:
On second thoughts, I'm not convinced this will be a net win, rather than just churn. `Con (ExplicitDict ...)` is a one-of hack to handle `run_accum`. But doing the same thing using `ExplicitMethods` is no less hacky. We'll have to add concrete syntax (it's not a regular primop), a custom inference rule, and we'll have to deal with `ExplicitMethods` appearing in pre-simplification IR, which we don't currently. So I think it's just trading one hack for another. If we're worried about the confusion of the name similarity, we could rename `Con (ExplicitDict ..)` to `Con (ExplicitAccumBaseMonoidDict _)`.

Reply:
Ok, fine.